services/grpcService.ts

```diff
@@ -203,27 +212,41 @@ export class GrpcBlockchainClient implements BlockchainClient {
     return details
   };
 
-  public async subscribeTransactions (
-    addresses: string[],
-    onTransactionNotification: (txn: GrpcTransaction.AsObject) => any,
-    onMempoolTransactionNotification: (txn: GrpcTransaction.AsObject) => any,
-    networkSlug: string
-  ): Promise<void> {
-    const createTxnStream = async (): Promise<void> => {
-      const client = this.getClientForNetworkSlug(networkSlug)
+  public async subscribeAddressesAddTransactions (addresses: Address[]): Promise<void> {
+    if (addresses.length === 0) return
+
+    const addressesAlreadySubscribed = addresses.filter(address => Object.keys(this.subscribedAddresses).includes(address.address))
+    if (addressesAlreadySubscribed.length === addresses.length) return
+    addressesAlreadySubscribed.forEach(address => {
+      console.log(`This address was already subscribed: ${address.address}`)
+    })
+    addresses = addresses.filter(address => !addressesAlreadySubscribed.includes(address))
+
+    const addressesInClients: KeyValueT<Address[]> = {}
+    this.availableNetworks.forEach(networkSlug => {
+      addressesInClients[networkSlug] = []
+    })
+    addresses.forEach(address => {
+      const prefix = getAddressPrefix(address.address)
+      if (!Object.keys(addressesInClients).includes(prefix)) { throw new Error(RESPONSE_MESSAGES.INVALID_ADDRESS_400.message) }
+      addressesInClients[prefix].push(address)
+    })
+
+    for (const [key, value] of Object.entries(addressesInClients)) {
+      const client = getGrpcClients()[key]
+      const addressesToSubscribe = value.map(address => address.address)
       const confirmedStream = await client.subscribeTransactions({
         includeMempoolAcceptance: false,
         includeBlockAcceptance: true,
-        addresses
+        addresses: addressesToSubscribe
       })
       const unconfirmedStream = await client.subscribeTransactions({
         includeMempoolAcceptance: true,
         includeBlockAcceptance: false,
-        addresses
+        addresses: addressesToSubscribe
       })
 
       const nowDateString = (new Date()).toISOString()
-
       // output for end, error or close of stream
       void confirmedStream.on('end', () => {
         console.log(`${nowDateString}: addresses ${addresses.join(', ')} confirmed stream ended`)
```

Comment:
I feel that here and below it shouldn't be `addresses.join` but `addressesToSubscribe.join`.

Reply:
indeed! done